libsolidity/codegen/ABIFunctions.cpp

```diff
@@ -1557,6 +1592,85 @@ string ABIFunctions::abiDecodingFunctionFunctionType(FunctionType const& _type,
 	});
 }
 
+string ABIFunctions::calldataAccessFunction(Type const& _type)
+{
+	string functionName = "calldata_access_" + _type.identifier();
+	return createFunction(functionName, [&]() {
+		if (_type.isDynamicallyEncoded())
+		{
+			unsigned int baseEncodedSize = _type.calldataEncodedSize();
+			solAssert(baseEncodedSize > 1, "");
+			Whiskers w(R"(
+				function <functionName>(base_ref, ptr) -> <return> {
+					let rel_offset_of_tail := calldataload(ptr)
+					if iszero(slt(rel_offset_of_tail, sub(sub(calldatasize(), base_ref), sub(<neededLength>, 1)))) { revert(0, 0) }
+					value := add(rel_offset_of_tail, base_ref)
+					<handleLength>
+				}
+			)");
+			if (_type.isDynamicallySized())
+			{
+				auto const* arrayType = dynamic_cast<ArrayType const*>(&_type);
+				solAssert(!!arrayType, "");
+				unsigned int calldataStride = arrayType->calldataStride();
+				w("handleLength", Whiskers(R"(
+					length := calldataload(value)
+					value := add(value, 0x20)
+					if gt(length, 0xffffffffffffffff) { revert(0, 0) }
+					if sgt(base_ref, sub(calldatasize(), mul(length, <calldataStride>))) { revert(0, 0) }
+				)")("calldataStride", toCompactHexWithPrefix(calldataStride)).render());
+				w("return", "value, length");
+			}
+			else
+			{
+				w("handleLength", "");
+				w("return", "value");
+			}
+			w("neededLength", toCompactHexWithPrefix(baseEncodedSize));
+			w("functionName", functionName);
+			return w.render();
+		}
+		else if (_type.isValueType())
+		{
+			// TODO: should this use the decoding function instead of cleaning directly?
+			//       What would such a call pass as "end"?
+			string cleanup;
+			if (auto const* functionType = dynamic_cast<FunctionType const*>(&_type))
+			{
+				// Should this be handled by cleanupFunction()?
```

Comment:
Yes, it should.

Reply:
My guess is that ``cleanupFunction`` didn't handle this before in order to make sure that one thinks about combined vs. two slot function types... I could make ``cleanupFunction`` handle this, but should it then take ``EncodingOptions`` in order to determine how to handle functions? In ``calldataAccessFunction`` it will always be combined, single-slot function pointers...